README.md

## compile:    gcc lis.c -o lis
## run:        ./lis [args]


## Usage: lis [OPTION]... [FILE]...
List information about the FILEs (the current directory by default).
Sort entries alphabetically if none of -cftuvSUX nor --sort is specified.
Implemented w.r.t. ls (GNU coreutils) 8.30

PASS_0 -> [struct beta] -> PASS_1 -> [flagSet] -> PASS_2




## PASS_0 (PARSING):
"lis_cmdParser.c"

The arguments of lis are parsed here.
If invalid arguments, it gives error message,
Otherwise, valid arguments filled in struct beta and passed to PASS_1.

struct beta contents, filled by command parser.
one node for one option, len=no. of args

description          |   flag  indOpt  argStr  coeff   num2   num1
---------------------+---------------------------------------------------------------------
normal argument      |   0      -      yes       -      -     -
sort option          |   1      -      yes       -      -     width/tabsize/PATRN
long option no args  |   2     ind      -        -      -     other/help/version
long option PATTERN  |   3     ind     yes       -      -     hide/ignore
long option COLS     |   4     ind      -        -      -     num
long option SIZE     |   5     ind      -      coeff   exp    base
long option WORD     |   6     ind      -        -      -     ind
long option +FORMAT  |   7     ind     yes       -      -     -



## PASS_1 (FLAGING):
"lis_pass1.c"

Reason of pass_1
    1. less/no strcmp.
    2. common flagSet(flaged by shiphen/dhiphen), like. if(char == 'a' || str == "all").
    3. reduce duplication, like. "-l" and "--format=long" both are similar(opt with arg).

According to struct beta, flagSet is filled by pass_1 using charSet.
The files "lis_set_flagSet.txt" and "lis_set_charSet.txt" have more information about flagSet and charSet respectively.



## PASS_2 (no name yet):
"lis_pass2.c"

FILTER_0 = INITIATING
    If help/version exist, print it and stop further processing.

FILTER_1 = RESHAPING
    Discard backup, hidden and pattern matched files (if =PATTERN).
    patternChecker()
        patternMatcher()

FILTER_2 = APPENDING
    fillSet()
        find file inforamtions (if required).

FILTER_3 = SEQUENCING
    sortSet()
        according to sort type
        sort type = name, mtime, ctime, atime, size, version, ext, dir_first
    reverseSet()

FILTER_4 = FORMATING
    format_size()
        sizeAsBlockSizeForm()
        sizeAsHumanReadable()
    format_quotIndcat()
        quoting to names and indicator to directories.
    format_timeStyle()
        full-iso, long-iso, locale, iso.

FILTER_5 = FLOWING
    horizntlPrintFlow()
        print by rows.
        start checking from N lines to 1 line.
    verticalPrintFlow()
        default, print by columns.
        start checking from 1 line to N lines.

FILTER_6 = PRINTING
    Printf contents according to flagSet.
    Random colours with option --color=auto(reduce vision fluctuation).

FILTER_7 = REMOVING
    Remove indicator, quoting from filename(if attached).

FILTER_8 = FREEING
    free memory malloc'd by pass2.


## SIZE_CALCULATION:
{for -s}
    normal:         size as it is.
    block_size:     upperBound(size/blkSize)
    human_readable:     recDiv(roundUp(size/base))

{for -l}
                alpha = upperBound(size/4096)*4
    normal:         alpha as it is.
                beta = alpha*1024
    block_size:     upperBound(beta/blkSize)
    human_readable:     recDiv(roundUp(beta/base))

base        1000/1024 according to --si/-h
recDiv(a/b) recurssive division while a>=b
roundUp     round upto 1 decimal digit.
        after recDiv if ans >= 10, then take upperBound



## REFERENCES-
LIBRARY:(full description at linux manual)
    stdio.h             printf()
    stdlib.h            malloc(), free(), exit()
    string.h            strcpy(), strcat(), strcmp(), strlen()
    dirent.h    struct dirent   opendir(), closedir()
    sys/ioctl.h struct winsize  ioctl()
    sys/stat.h  struct stat stat()
    pwd.h       struct passwd   getpwuid()
    grp.h       struct group    getgrgid()
    time.h      struct tm   gmtime(), strftime()


##########################################
'lis' skips some options of 'ls' as below:
    INCOMPLETE_WITH_FLAGSET_
        --hyperlink
    -Z  --context
    -b  --escape, --quoting-style=escape
    -T  --tabsize
    -W  --width

    INCOMPLETE_WITHOUT_FLAGSET_
    -D  --dired
    -L  --dereference
    -H  --dereference-command-line
        --dereference-command-line-symlink-to-dir


## FOR USER:
    Sometimes the output is correct but useless or unexpected.
    So please report as a bug, Mistakes ready to improve.


## INPUTS AND OUTPUTS:
Only valid options and their arguments are worth for further processing,
    e.g. "--all -a" are valid
         "--za -z" are exited with error message.

lis follows smooth input strategy, i.e. Rule of last action have higher priority,
    e.g. "-aA", "a" is desabled, output is according to rule of "A".
         "--blocksize=2" desable without "-l".
         "-m -l", output is on newline for each entry rather than comma seperated.
         "--" gives list of all long options.
         "--f" print possibilities as '--file-type' '--format' '--full-time'.
         two others are as below.

For options and their arguments, substring is allowed, except ambiguous options,
    e.g. "--fo=lo" considered as "--format=long".

Arguments of option allows synonyms,
    e.g. "--fo=verbose" "--fo=long" both are similer.

Number and unit in argument must be in proper format and unit is within range,
    e.g. "--width=23 --blo=12M -w23" are valid.
         "--blo=1r --wi=f -w -wa" are invalid.
         "--blo=123e" too large, (max valid value is 2^64-1 = 18,446,744,073,709,551,615)

Pattern have only single *, before or after, Also in proper format,
    e.g. "--hide=abc --ignore=*txt -Iabc*" are valid.
         "--hide=ab*cd" misbehave.
         "-I" without argument.

## Output:
    By default, follows vertical print flow.
    Directories are in blue colour, files are in white.
    With "--color=auto", directories with blue background, files with random colours.

Use zoom out(smaller font size) in "lis_set_flagSet.txt" and "lis_set_charSet.txt" files,
if content is exceeding line(May be it misunderstandable).



about/HISTORY.md


## Timeline : Mca 4th Semester
## Guide : Prof. Achyut Roy
## Subjects : 
##     CS-402 Software Development - 1
##     CS-403 Software Development - 2


23/03/2020	start
06/05/2020	complete


lis.c

#include<stdio.h>
#include<stdlib.h>
#include<string.h>

#include<dirent.h>	//struct dirent, opendir(), closedir()
#include<sys/ioctl.h>	//struct winsize, ioctl()

#include"lis_cmdParser.c"
#include"lis_pass1.c"
#include"lis_pass2.c"

void pass2_rec(int, int*, char*(*)[], char*, int, int);
void createTempSet(struct set*);
void createSet(char*, struct set*);

int main(int argc, char *argv[]) {
	int ind = 0;
	int errStat = 0;

	//PASS_0
	argv[0] = (char*)malloc(3+1);
	strcpy(argv[0], "lis");
	struct beta beta1[argc]; //[0] is fixed for help/version
	//initiating beta
	ind = -1;
	while(++ind < argc) {
		beta1[ind].flag = -1;
		beta1[ind].indOpt = -1;
		beta1[ind].num1 = -1;
		beta1[ind].num2 = -1;
		beta1[ind].coeff = -1;
		beta1[ind].str = NULL;
	}

	//argument parsing
	errStat = parseCommand(argc, argv, beta1);
	if(errStat == -1) {
		//printf("serious trouble with exit status: 2\n");
		exit(0);
	}

/*	//###########################################################
	//printing struct beta, filled by command parser
	ind = -1;
	printf("list of options:\n");
	while(++ind < argc) {
		if(beta1[ind].flag != -1)
			printf("%d\t", beta1[ind].flag);
		if(beta1[ind].indOpt != -1)
			printf("%d\t", beta1[ind].indOpt);
		if(beta1[ind].coeff != -1)
			printf("%lu\t", beta1[ind].coeff);
		if(beta1[ind].num1 != -1)
			printf("%d\t", beta1[ind].num1);
		if(beta1[ind].num2 != -1)
			printf("%d\t", beta1[ind].num2);
		if(beta1[ind].str != NULL)
			printf("\"%s\"", beta1[ind].str);
		printf("\n");
		if(beta1[0].flag != -1)
			break;
	}
	printf("\n");
*/	//###########################################################



	//PASS_1
	char *pattern[2][argc]; //[0][0] time-style, [0][>0] hide, [1][] ignore
	pattern[0][0] = NULL;
	pattern[0][1] = NULL;
	pattern[1][1] = NULL;

	char *pathList[argc];
	pathList[0] = (char*)malloc(3); //+1 for hidden '\'
	strcpy(pathList[0], ".");
	pathList[1] = NULL;

	int flagSet[42] = {
		0,				//help/version
		1,1,0,0,			//file type
		0,0,0,
		0,0,0,
		0,0,0,0,0,0,0,0,0,0,0,5,0,	//file info
		1,1,1,1,			//block size 
		0,				//size flag
		0,8,				//width,tabsize
		1,0,0,				//sort
		0,0,0,1,2,
		0,0,0
	};

/*	//#####################################################################
	ind = -1;
	while(++ind < 41) printf("%4d", ind);
	printf("\n");
	ind = -1;
	while(++ind < 41) printf("%4d", flagSet[ind]);
	printf("\n");
*/	//#####################################################################

	errStat = pass1(argc, beta1, flagSet, pattern, pathList);

	//freeing struct beta melloc'd by command parser
	ind = -1;
	while(++ind < argc)
		if(beta1[ind].str != NULL)
			free(beta1[ind].str);

/*	//#####################################################################
	ind = -1;
	while(++ind < 41) printf("%4d", flagSet[ind]);
	printf("\n");
	pass1print(argc, flagSet, pattern, pathList);
*/	//#####################################################################



	//PASS_3
	//for each path given by command line arguments
	struct winsize ws;
	ioctl(0, TIOCGWINSZ, &ws);
	int winWidth = ws.ws_col;
	int recFlag = flagSet[5];
	ind = -1;
	while(pathList[++ind] != NULL) {
		pass2_rec(argc, flagSet, pattern, pathList[ind], recFlag, winWidth);
		free(pathList[ind]);
		pathList[ind] = NULL;
	}

}


void pass2_rec(int argc, int *flagSet, char *pattern[2][argc], char *path, int recFlag, int winWidth) {
	struct set *keySet = (struct set*)malloc(sizeof(struct set));
	struct set *nextSet, *tempSet;

	createSet(path, keySet); //build a link list(keySet) of file/dir of "path"
	pass2(argc, flagSet, pattern, path, keySet, winWidth);

	if(recFlag == 1) {
		nextSet = keySet;
		while(nextSet->link != NULL) {
			char *pathStr = (char*)malloc(strlen(path)+1+strlen(nextSet->link->strName)+1);
			strcpy(pathStr, path);
			strcat(pathStr, "/");
			strcat(pathStr, nextSet->link->strName);

			DIR *dirPtr = opendir(pathStr);
			if(dirPtr == NULL || !strcmp(nextSet->link->strName, ".") || !strcmp(nextSet->link->strName, "..")) {
				//freeing keySet(files) melloc'd by pass2
				tempSet = nextSet->link;
				nextSet->link = nextSet->link->link;
				free(tempSet->strName);
				tempSet->strName = NULL;
				tempSet->link = NULL;
				free(tempSet);
				tempSet = NULL;
			}
			else { //remaining are directories
				free(nextSet->link->strName);
				nextSet->link->strName = (char*)malloc(strlen(pathStr)+1);
				strcpy(nextSet->link->strName, pathStr);
				nextSet = nextSet->link;
			}
			closedir(dirPtr);
			free(pathStr);
		}
	}

	//do recurssion and freeing keySet(dir) melloc'd by pass2
	while(keySet->link != NULL) {
		if(recFlag == 1)
			pass2_rec(argc, flagSet, pattern, keySet->link->strName, recFlag, winWidth);

		//freeing keySet(files) melloc'd by pass2
		tempSet = keySet->link;
		keySet->link = keySet->link->link;
		free(tempSet->strName);
		tempSet->strName = NULL;
		tempSet->link = NULL;
		free(tempSet);
		tempSet = NULL;

	}
	free(keySet);
	keySet = NULL;
}


void createSet(char *path, struct set *keySet) {
	struct dirent *dirent1;
	DIR *dirPtr = opendir(path);

	if(dirPtr == NULL) { //lis command line argument is file
		keySet->link = (struct set*)malloc(sizeof(struct set));
		keySet = keySet->link;

		char *fileName = NULL;
		int len = strlen(path);
		int i = len-1;
		while(1) {
			if(i == 0) {
				fileName = (char*)malloc(strlen(path)+1);
				strcpy(fileName, path);
				free(path);
				path = (char*)malloc(3);
				strcpy(path, ".");
				break;
			}
			if(path[i] == '/') {
				path[i] = '\0';
				fileName = (char*)malloc(strlen(path+i+1)+1);
				strcpy(fileName, path+i+1);
				break;
			}
			i--;
		}

		initSet(keySet);
		keySet->strName = (char*)malloc(strlen(fileName)+1);
		strcpy(keySet->strName, fileName);
		free(fileName);
	}
	else {
		while(1) { //dir
			dirent1 = readdir(dirPtr);
			if(dirent1 == NULL) break;

			keySet->link = (struct set*)malloc(sizeof(struct set));
			keySet = keySet->link;
			initSet(keySet);
			keySet->strName = (char*)malloc(strlen(dirent1->d_name)+1);
			strcpy(keySet->strName, dirent1->d_name);
		}
	}
	closedir(dirPtr);
}



lis_pass1.c

int pass1(int, struct beta*, int*, char*(*)[], char**);
void flagSetter(char*, int*);
void pass1print(int, int*, char*(*)[], char**);

int pass1(int argc, struct beta beta1[], int flagSet[], char *pattern[2][argc], char *pathList[]) {
	char *charSet[58] = {
		"C0","D1E0","D1E1","F1","G1","H1","I1","J1",
		"K1L1R1S0T1U1V1W5c2","K1L1R1S1T0U1V1W5c2","K1L1R1S1T1U1V1W1c2",
		"K1L1R1S1T1U1V1W5c2","K1L1R1S2T2U1V1W5c2",
		"M1","O1","S0","V2","V3","W1","W2","W3","W4","W5","X2","]1","]2","]3","`0",
		"`0D1E1i0","`2","`2a1b3","`3","`4a1b3","`5","`6","`7","`8",
		"c1K0L0R0S0T0U0V0W0","c2","d0",
		"d1","d2","d3","e0f0","e0f1","e0f4","e1f1","e2f2","e3f3",
		"g1K0L0R0S0T0U0V0W0c0","g2K0L0R0S0T0U0V0W0c0","h0","h1","h2","i0","i1","i2","j0"
	};

	int shiphen[74] = {
		38,
		-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,
		1,0,50,57/**/,-1,42,15,57/**/,-1,-1,-1,57/**/,-1,43,-1,-1,47,3,33,-1,27,-1,-1,35,-1,6,
		-1,-1,-1,-1,-1,-1,
		2,45,32,5,-1,28,8,26,14,-1,23,11,37,12,9,40,44,4,7,31,30,34,-1,49,-1,-1
	};

	//dhiphen+shiphen	only dhiphen	dhiphen+shiphen	indBig_of_dhiphen1
	int dhiphen[43] = {
		2,1,			13,24,		42,		0,
		6,57/**/,57/**/,	57/**/,		5,57/**/,45,
					41,				3,
					10,36,-1,-1,	44,26,		8,
		-1,0,							11,
		14,23,43,15,12,47,					15,
		3,4,			44,25,		7,		25,
					-1,				30,32,
					-1,-1
	};

	int dhiphen1[37] = {
		54,55,56,			//color
		11,37,49,50,38,			//format
		51,52,53,			//hyperlink
		39,40,41,42,			//ind style
		43,44,46,46,44,47,47,45,48,48,	//quot style
		27,29,33,35,34,			//sort
		17,16,				//time
		18,19,20,21,22			//time style
	};


	int ind = 0;
	int indPat[2] = {0, 0};
	int indPath = 0;
	flagSet[0] = beta1[0].num1; // 1-help 2-version

	if(flagSet[0] != 1 || flagSet[0] != 2)
	while(++ind < argc) {
		if(beta1[ind].flag == 0) { // PATHLIST
			int len = strlen(beta1[ind].str);
			if(beta1[ind].str[len-1] == '/') {
				if(strlen(beta1[ind].str) == 1) //root directory
					flagSet[5] = 0; //desable recurssion
				else
					beta1[ind].str[len-1] = '\0';
			}
			else
				len++; //+1 for '\', if indicator required
			pathList[indPath] = (char*)malloc(len+1);
			strcpy(pathList[indPath], beta1[ind].str);
			pathList[indPath+1] = NULL;
			indPath++;
		}
		else if(beta1[ind].flag == 7) { // =+FORMAT
			flagSetter(charSet[dhiphen1[dhiphen[beta1[ind].indOpt]+beta1[ind].num1]], flagSet);
			pattern[0][0] = (char*)malloc(strlen(beta1[ind].str)+1);
			strcpy(pattern[0][0], beta1[ind].str);
		}
		else if(beta1[ind].flag == 3) { // =PATTERN
			pattern[beta1[ind].num1][++indPat[beta1[ind].num1]] = (char*)malloc(strlen(beta1[ind].str)+1);
			strcpy(pattern[beta1[ind].num1][indPat[beta1[ind].num1]], beta1[ind].str);
		}
		else if(beta1[ind].flag == 4) { // =COLS
			flagSet[29+(beta1[ind].indOpt/37)] = beta1[ind].num1;
		}
		else if(beta1[ind].flag == 5) { // =SIZE
			flagSetter(charSet[24], flagSet);

			flagSet[26] = beta1[ind].num1;
			flagSet[27] = beta1[ind].num2;

			unsigned long long1 = beta1[ind].coeff;
			flagSet[24] = long1;
			long1 = long1 >> (sizeof(long1)*4);
			flagSet[25] = long1;
		}
		else if(beta1[ind].flag == 1) { // SHIPHEN
			int len = strlen(beta1[ind].str);

			if(beta1[ind].num1 == 2) { //tabsize
				flagSet[30] = beta1[ind].num2;
				continue;
			}
			if(beta1[ind].num1 == 1) { //width
				flagSet[29] = beta1[ind].num2;
				continue;
			}
			if(beta1[ind].num1 == 3) { //ignore
				pattern[1][++indPat[1]] = (char*)malloc(strlen(beta1[ind].str+beta1[ind].num2)+1);
				strcpy(pattern[1][indPat[1]], beta1[ind].str+beta1[ind].num2);
				continue;
			}

			int ind1 = -1;
			while(++ind1 < len) {
					flagSetter(charSet[shiphen[beta1[ind].str[ind1]-49]], flagSet);
			}
		}
		else if(beta1[ind].flag == 2) { // =NULL
			flagSetter(charSet[dhiphen[beta1[ind].indOpt]], flagSet);
		}
		else if(beta1[ind].flag == 6) { // =WORD/[=WHEN]
			flagSetter(charSet[dhiphen1[dhiphen[beta1[ind].indOpt]+beta1[ind].num1]], flagSet);
		}
	}
	pattern[0][++indPat[0]] = NULL;
	pattern[1][++indPat[1]] = NULL;
	return indPath;
}


void flagSetter(char *charSetStr, int flagSet[]) {
	int ind = 0;
	while(charSetStr[ind] != '\0') {
		flagSet[charSetStr[ind]-65] = charSetStr[ind+1]-48;
		ind += 2;
	}
}


void pass1print(int argc, int flagSet[], char *pattern[2][argc], char *pathList[]) {
	int ind;
	if(flagSet[0] == 1) {
		printf("help file.\n");
		return;
	}
	if(flagSet[0] == 2) {
		printf("version file.\n");
		return;
	}

	ind = -1;
	printf("\n\npathList:\t");
	while(pathList[++ind] != NULL) printf("%s ", pathList[ind]);

	printf("\ntimeStyle:\t");
	if(pattern[0][0] != NULL) printf("%s", pattern[0][0]);

	ind = 0;
	printf("\nhide:\t\t");
	while(pattern[0][++ind] != NULL) printf("%s ", pattern[0][ind]);

	ind = 0;
	printf("\nignore:\t\t");
	while(pattern[1][++ind] != NULL) printf("%s ", pattern[1][ind]);
	printf("\n\n");

	if(flagSet[1] != 0) printf("normal files\n");
	if(flagSet[2] != 0) printf("backup files\n");
	if(flagSet[3] != 0) printf("hidden files\n");
	if(flagSet[4] != 0) printf("dot files\n");
	printf("\n");
	if(flagSet[5] != 0) printf("recursion\n");
	if(flagSet[6] != 0) printf("reverse\n");
	if(flagSet[7] != 0) printf("directory\n");
	printf("\n");
	if(flagSet[8] != 0) printf("context\n");
	if(flagSet[9] != 0) printf("-s\n");
	if(flagSet[10] != 0) printf("-l\n");
	printf("\n");
	if(flagSet[11] != 0) printf("permision\n");
	if(flagSet[12] != 0) printf("author\n");
	if(flagSet[13] != 0) printf("mode\n");
	if(flagSet[14] != 0) printf("inode\n");
	if(flagSet[15] != 0) printf("dev\n");
	if(flagSet[16] != 0) printf("rdev\n");
	if(flagSet[17] != 0) printf("hlink\n");
	if(flagSet[18] != 0) printf("uid\n");
	if(flagSet[19] != 0) printf("gid\n");
	if(flagSet[20] != 0) printf("     size:\t%d\n", flagSet[20]);
	if(flagSet[21] != 0) printf("     time:\t%d\n", flagSet[21]);
	if(flagSet[22] != 0) printf("timeStyle:\t%d\n", flagSet[22]);
	if(flagSet[23] != 0) printf("   blocks:\t%d\n", flagSet[23]);
	printf("\n");
	if(flagSet[24] != 0) printf("    coeff:\t%d\n", flagSet[24]);
	if(flagSet[25] != 0) printf("    coeff:\t%d\n", flagSet[25]);
	if(flagSet[26] != 0) printf("     base:\t%d\n", flagSet[26]);
	if(flagSet[27] != 0) printf("      exp:\t%d\n", flagSet[27]);
	printf("\n");
	if(flagSet[28] != 0) printf("  hr-base:\t%d\n", flagSet[28]);
	printf("\n");
	if(flagSet[29] != 0) printf("    width:\t%d\n", flagSet[29]);
	if(flagSet[30] != 0) printf("  tabsize:\t%d\n", flagSet[30]);
	printf("\n");
	if(flagSet[31] != 0) printf("     sort:\t%d\n", flagSet[31]);
	if(flagSet[32] != 0) printf("  sort -l:\t%d\n", flagSet[32]);
	if(flagSet[33] != 0) printf(" sort -lt:\t%d\n", flagSet[33]);
	printf("\n");
	if(flagSet[34] != 0) printf("seperator:\t%d\n", flagSet[34]);
	if(flagSet[35] != 0) printf("indicator:\t%d\n", flagSet[35]);
	if(flagSet[36] != 0) printf("     quot:\t%d\n", flagSet[36]);
	if(flagSet[37] != 0) printf("  esc-seq:\t%d\n", flagSet[37]);
	if(flagSet[38] != 0) printf("     flow:\t%d\n", flagSet[38]);
	printf("\n");
	if(flagSet[39] != 0) printf("hyperlink:\t%d\n", flagSet[39]);
	if(flagSet[40] != 0) printf("    color:\t%d\n", flagSet[40]);
	printf("\n");
}



imFileParser/imFileReader.c
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<unistd.h>
#include<fcntl.h>

struct blkOptions {
	int shiphenFlagBuff[64];

	int lenOptAll;
	char **allOptions;   // [lenOptAll]

	int lenOpts[4];
	int *listOpts[4];
	
	int numArgsList;
	int *lenArgs;        // [numArgsList]
	char ***listArgs;    // [numArgsList]
};

int readPrerequisiteFile(char*, struct blkOptions*);
void printBlkOptions(struct blkOptions);
int readLine(int, char*);
int readNumber(int);

/*
int main() {
	struct blkOptions blkOpt;
	char *fileName = (char*)malloc(14);
	strcpy(fileName, "imFile_lis.txt");
	int errStat = readPrerequisiteFile(fileName, &blkOpt);
	if(errStat == -1)
		exit(0);
	printBlkOptions(blkOpt);
	free(fileName);
	exit(0);
}
*/

int readPrerequisiteFile(char *fileName, struct blkOptions *blkOpt) {
	char lineBuff[100];
	int lineLen;
	int num, i;

	int fdOpen = open(fileName, O_RDONLY);
	if(fdOpen == 0) {
		printf("prerequisite file '%s' is not opened\n", fileName);
		return -1;
	}

	//init shiphenFlagBuff[]
	i = -1;
	while(++i < 64)
		blkOpt->shiphenFlagBuff[i] = 0;

	//read shiphenFlagBuff[]
	i = -1;
	lineLen = readLine(fdOpen, lineBuff);
	while(++i < lineLen) {
		if(lineBuff[i] == '1')
			lineBuff[i] = 64;
		blkOpt->shiphenFlagBuff[lineBuff[i]-64] = 1;
	}

	//read lenOptAll, allOptions[]
	i = -1;
	num = readNumber(fdOpen);
	blkOpt->lenOptAll = num;
	blkOpt->allOptions = (char**)malloc(sizeof(char**)*num);
	while(++i < num) {
		lineLen = readLine(fdOpen, lineBuff);
		blkOpt->allOptions[i] = (char*)malloc(lineLen+1);
		strcpy(blkOpt->allOptions[i], lineBuff);
	}

	//read lenOpts[], listOpts[]
	i = -1;
	while(++i < 4) {
		num = readNumber(fdOpen);
		blkOpt->lenOpts[i] = num;
		blkOpt->listOpts[i] = (int*)malloc(sizeof(int*)*num);
		int j = -1;
		while(++j < num)
			blkOpt->listOpts[i][j] = readNumber(fdOpen);
	}

	// read numArgs, lenArgs[], listArgs[]
	i = -1;
	blkOpt->numArgsList = num;
	blkOpt->lenArgs = (int*)malloc(sizeof(int*)*num);
	blkOpt->listArgs = (char***)malloc(sizeof(char***)*num);
	while(++i < num) {
		int j = -1;
		int num1 = readNumber(fdOpen);
		blkOpt->lenArgs[i] = num1;
		blkOpt->listArgs[i] = (char**)malloc(sizeof(char**)*num1);
		while(++j < num1) {
			lineLen = readLine(fdOpen, lineBuff);
			blkOpt->listArgs[i][j] = (char*)malloc(sizeof(char*)*lineLen);
			strcpy(blkOpt->listArgs[i][j], lineBuff);
		}
	}

	return 0;
}

void printBlkOptions(struct blkOptions blkOpt) {
	int i, num;

	//print shiphenFlagBuff[64]
	i = -1;
	printf("shiphenFlagBuff[64] = {");
	while(++i < 64-1) {
		if(i%20 == 0) printf("\n\t");
		if(i%10 == 0) printf("  ");
		printf("%2d, ", blkOpt.shiphenFlagBuff[i]);
	}
	printf("%2d}", blkOpt.shiphenFlagBuff[i]);

	//print lenOptAll, allOptions[]
	i = -1;
	printf("\n\n");
	num = blkOpt.lenOptAll;
	printf("lenOptAll = %d\n", num);
	printf("allOptions[%d] = {", num);
	while(++i < num-1)
		printf("\"%s\", ", blkOpt.allOptions[i]);
	printf("\"%s\"}", blkOpt.allOptions[i]);

	//print lenOpts[], listOpts[]
	i = -1;
	printf("\n");
	char *optWithArg[4] = {"Word","Cols","Size","Ptrn"};
	while(++i < 4) {
		num = blkOpt.lenOpts[i];
		printf("\noptWithArg%s[%d][%d] = {", optWithArg[i], i, num);
		int j = -1;
		while(++j < num-1)
			printf("%d, ", blkOpt.listOpts[i][j]);
		printf("%d}", blkOpt.listOpts[i][j]);
	}
	
	//print numArgsList, lenArgs, listArgs[]
	i = -1;
	printf("\n\n");
	num = blkOpt.numArgsList;
	printf("numArgsList = %d\n", num);
	while(++i < num) {
		printf("argList[%d][%d] = {", i, blkOpt.lenArgs[i]);
		int j = -1;
		while(++j < blkOpt.lenArgs[i]-1)
			printf("\"%s\", ", blkOpt.listArgs[i][j]);
		printf("\"%s\"}", blkOpt.listArgs[i][j]);
		printf("\n");
	}
	printf("\n");

}

int readLine(int fdOpen, char *lineBuff) {
	size_t cin;
	char rdBuff;
	int lineLen = 0;

	while(1) {
		cin = read(fdOpen, &rdBuff, 1);
		if(cin == 0 || rdBuff == '\n') {
			lineBuff[lineLen] = '\0';
			return lineLen;
		}
		lineBuff[lineLen] = rdBuff;
		lineLen++;
	}
}

int readNumber(int fdOpen) {
	size_t cin;
	char rdBuff;
	int num = 0;

	while(1) {
		cin = read(fdOpen, &rdBuff, 1);
		if(rdBuff >= 48 && rdBuff <= 57)
			num = num*10+(rdBuff-48);
		else
			return num;

	}
}



lis_print.c

void printVersion() {
printf("lis version 1.0\n");
printf("Implemented w.r.t. ls (GNU coreutils) 8.30\n");
printf("Copyright (C) under PUCSD Assignments of Software Development\n");
printf("\n");
printf("Lead by Prof. Achyut K. Roy.\n");
printf("Written by Ashok J. Kumawat.\n");
}


void printHelp() {
printf("Usage: lis [OPTION]... [FILE]...\n");
printf("List information about the FILEs (the current directory by default).\n");
printf("Sort entries alphabetically if none of -cftuvSUX nor --sort is specified.\n");
printf("\n");
printf("Mandatory arguments to long options are mandatory for short options too.\n");
printf("  -a, --all                  do not ignore entries starting with .\n");
printf("  -A, --almost-all           do not list implied . and ..\n");
printf("      --author               with -l, print the author of each file\n");
printf("      --block-size=SIZE      with -l, scale sizes by SIZE when printing them;\n");
printf("                               e.g., '--block-size=M'; see SIZE format below\n");
printf("  -B, --ignore-backups       do not list implied entries ending with ~\n");
printf("  -c                         with -lt: sort by, and show, ctime (time of last\n");
printf("                               modification of file status information);\n");
printf("                               with -l: show ctime and sort by name;\n");
printf("                               otherwise: sort by ctime, newest first\n");
printf("  -C                         list entries by columns\n");
printf("      --color[=WHEN]         colorize the output; WHEN can be 'always' (default\n");
printf("                               if omitted), 'auto', or 'never'; more info below\n");
printf("  -d, --directory            list directories themselves, not their contents\n");
printf("  -f                         do not sort, enable -aU, disable -ls --color\n");
printf("  -F, --classify             append indicator (one of */=>@|) to entries\n");
printf("      --file-type            likewise, except do not append '*'\n");
printf("      --format=WORD          across -x, commas -m, horizontal -x, long -l,\n");
printf("                               single-column -1, verbose -l, vertical -C\n");
printf("      --full-time            like -l --time-style=full-iso\n");
printf("  -g                         like -l, but do not list owner\n");
printf("      --group-directories-first\n");
printf("                             group directories before files;\n");
printf("                               can be augmented with a --sort option, but any\n");
printf("                               use of --sort=none (-U) disables grouping\n");
printf("  -G, --no-group             in a long listing, don't print group names\n");
printf("  -h, --human-readable       with -l and -s, print sizes like 1K 234M 2G etc.\n");
printf("      --si                   likewise, but use powers of 1000 not 1024\n");
printf("      --hide=PATTERN         do not list implied entries matching shell PATTERN\n");
printf("                               (overridden by -a or -A)\n");
printf("      --indicator-style=WORD  append indicator with style WORD to entry names:\n");
printf("                               none (default), slash (-p),\n");
printf("                               file-type (--file-type), classify (-F)\n");
printf("  -i, --inode                print the index number of each file\n");
printf("  -I, --ignore=PATTERN       do not list implied entries matching shell PATTERN\n");
printf("  -k, --kibibytes            default to 1024-byte blocks for disk usage;\n");
printf("                               used only with -s and per directory totals\n");
printf("  -l                         use a long listing format\n");
printf("  -m                         fill width with a comma separated list of entries\n");
printf("  -n, --numeric-uid-gid      like -l, but list numeric user and group IDs\n");
printf("  -N, --literal              print entry names without quoting\n");
printf("  -o                         like -l, but do not list group information\n");
printf("  -p, --indicator-style=slash\n");
printf("                             append / indicator to directories\n");
printf("  -q, --hide-control-chars   print ? instead of nongraphic characters\n");
printf("      --show-control-chars   show nongraphic characters as-is (the default,\n");
printf("                               unless program is 'ls' and output is a terminal)\n");
printf("  -Q, --quote-name           enclose entry names in double quotes\n");
printf("      --quoting-style=WORD   use quoting style WORD for entry names:\n");
printf("                               literal, locale, shell, shell-always,\n");
printf("                               shell-escape, shell-escape-always, c, escape\n");
printf("                               (overrides QUOTING_STYLE environment variable)\n");
printf("  -r, --reverse              reverse order while sorting\n");
printf("  -R, --recursive            list subdirectories recursively\n");
printf("  -s, --size                 print the allocated size of each file, in blocks\n");
printf("  -S                         sort by file size, largest first\n");
printf("      --sort=WORD            sort by WORD instead of name: none (-U), size (-S),\n");
printf("                               time (-t), version (-v), extension (-X)\n");
printf("      --time=WORD            with -l, show time as WORD instead of default\n");
printf("                               modification time: atime or access or use (-u);\n");
printf("                               ctime or status (-c); also use specified time\n");
printf("                               as sort key if --sort=time (newest first)\n");
printf("      --time-style=TIME_STYLE  time/date format with -l; see TIME_STYLE below\n");
printf("  -t                         sort by modification time, newest first\n");
printf("  -u                         with -lt: sort by, and show, access time;\n");
printf("                               with -l: show access time and sort by name;\n");
printf("                               otherwise: sort by access time, newest first\n");
printf("  -U                         do not sort; list entries in directory order\n");
printf("  -v                         natural sort of (version) numbers within text\n");
printf("  -x                         list entries by lines instead of by columns\n");
printf("  -X                         sort alphabetically by entry extension\n");
printf("  -1                         list one file per line.  Avoid '\n' with -q or -b\n");
printf("      --help                 display this help and exit\n");
printf("      --version              output version information and exit\n");
printf("\n");
printf("The following options are usually not available to work\n");
printf("      --hyperlink[=WHEN]     hyperlink file names; WHEN can be 'always'\n");
printf("                               (default if omitted), 'auto', or 'never'\n");
printf("  -b, --escape               print C-style escapes for nongraphic characters\n");
printf("  -T, --tabsize=COLS         assume tab stops at each COLS instead of 8\n");
printf("  -w, --width=COLS           set output width to COLS.  0 means no limit\n");
printf("  -Z, --context              print any security context of each file\n");
printf("\n");
printf("The SIZE argument is an integer and optional unit (example: 10K is 10*1024).\n");
printf("Units are K,M,G,T,P,E,Z,Y (powers of 1024) or KB,MB,... (powers of 1000).\n");
printf("\n");
printf("The TIME_STYLE argument can be full-iso, long-iso, iso, locale, or +FORMAT.\n");
printf("FORMAT is interpreted like in date(1).\n");
printf("\n");
printf("Using color to distinguish file view is disabled both by default and\n");
printf("with --color=never.  With --color=auto, lis emits random colors for files.\n");
printf("\n");
printf("Exit status:\n");
printf(" 0  if OK,\n");
printf(" -1 if serious trouble (e.g., cannot access command-line argument).\n");
printf("\n");
}

